### PR TITLE
 DBZ-776 A first set of metrics implemented in new framework

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-776

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/pipeline/metrics/ChangeEventSourceMetricsMXBean.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Metrics that are common for both snapshot and streaming change event sources
+ *
+ * @author Jiri Pechanec
+ *
+ */
+public interface ChangeEventSourceMetricsMXBean {
+
+    String getLastEvent();
+    long getSecondsSinceLastEvent();
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/metrics/ChangeEventSourceMetricsMXBean.java]: Seconds seems a bit too coarse, as there can (and hopefully are) multiple events per second. How about changing it to milli-seconds at least?

File: debezium-core/src/main/java/io/debezium/pipeline/source/spi/ArrivedEventListener.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ */
+package io.debezium.pipeline.source.spi;
+
+/**
+ * A class invoked by {@link EventDispatcher} whenever an event is available for processing.
+ *
+ * @author Jiri Pechanec
+ *
+ */
+public interface ArrivedEventListener {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/source/spi/ArrivedEventListener.java]: Perhaps `DataChangeEventListener?

File: debezium-core/src/main/java/io/debezium/pipeline/source/spi/SnapshotChangeEventSource.java [hunk 3]
```diff
@@ -27,4 +27,11 @@
      *             in case the snapshot was aborted before completion
      */
     SnapshotResult execute(ChangeEventSourceContext context) throws InterruptedException;
+
+    /**
+     * Set a listener that is invoked for important events that happens during snapshot process, like start, stop, table completed etc.
+     *
+     * @param listener
+     */
+    public void setSnapshotProgressListener(SnapshotProgressListener listener);
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/source/spi/SnapshotChangeEventSource.java]: Hum hum, I wished we could avoid this so the listener reference in the implementation can be final.

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java [hunk 4]
```diff
@@ -50,6 +51,7 @@
     private final DataCollectionFilter<T> filter;
     private final ChangeEventCreator changeEventCreator;
     private final Heartbeat heartbeat;
+    private ArrivedEventListener eventListener = ArrivedEventListener.NO_OP;
```
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java]: That was my thinking - I thought more in terms of OpenTracing but the possibilities are endless.

**Comment by jpechane** [comment]
@gunnarmorling All comments incorporated into the code

**Comment by gunnarmorling** [comment]
@jpechane Great work overall! I've pushed one more commit. With that it's good to go from my POV, I've also filed https://issues.jboss.org/browse/DBZ-978 to capture some follow-up tasks. Can you check out my last commit and merge this PR if you don't have any objections? Thanks!

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java [hunk 5]
```diff
@@ -105,12 +105,12 @@ public SnapshotReceiver getSnapshotChangeEventReceiver() {
      * {@link ChangeEventCreator} for converting them into data change events.
      */
     public void dispatchDataChangeEvent(T dataCollectionId, ChangeRecordEmitter changeRecordEmitter) throws InterruptedException {
+        eventListener.onEvent();
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java]: Ok, that's a nice idea. But wouldn't it make sense to only move it back when that separate counter is introduced? But I'm not hung up on this, feel free to adjust as you prefer and then merge.

[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java]: @gunnarmorling Well, we are still in beta so you are right, we can afford to it this way. Going to merge,

**Pull request merged**